### PR TITLE
- Create types for each field in EventTypes to prevent impossible entries

### PR DESCRIPTION
**Pull request by johnsonw**
- Create types for each field in EventTypes to prevent impossible entries

Signed-off-by: Will Johnson <william.c.johnson@intel.com>

### CODE REVIEW

File: IML/DeviceScannerDaemon/EventTypes.fs [hunk 1]
```diff
@@ -98,27 +128,32 @@ let private findOrFail (key:string) x =
     | Some(x) -> unwrapString x
     | None -> failwith (sprintf "Could not find key %s in %O" key x)
 
+let isOne = Option.map(function
```
Comment (jgrund) [inline IML/DeviceScannerDaemon/EventTypes.fs]: Can this be private?

File: IML/DeviceScannerDaemon/EventTypes.fs [hunk 2]
```diff
@@ -98,27 +128,32 @@ let private findOrFail (key:string) x =
...
-let private parseIdPartEntryNumber = findOrNone "ID_PART_ENTRY_NUMBER"
-let private parseImlSize = findOrNone "IML_SIZE"
-let private parseImlScsi80 = findOrNone "IML_SCSI_80"
-let private parseImlScsi83 = findOrNone "IML_SCSI_83"
-let private parseImlRo = findOrNone "IML_IS_RO"
+let private parseIdVendor x = findOrNone "ID_VENDOR" x |> Option.map IdVendor
+let private parseIdModel x = findOrNone "ID_MODEL" x |> Option.map IdModel
+let private parseIdSerial x = findOrNone "ID_SERIAL" x |> Option.map IdSerial
+let private parseIdFsType x = findOrNone "ID_FS_TYPE" x |> Option.map IdFsType
+let private parseIdPartEntryNumber x = findOrNone "ID_PART_ENTRY_NUMBER" x |> Option.map int |> Option.map IdPartEntryNumber
```
Comment (jgrund) [inline IML/DeviceScannerDaemon/EventTypes.fs]: Could be a compose here.

File: IML/DeviceScannerDaemon/EventTypes.fs [hunk 3]
```diff
@@ -193,5 +220,5 @@ let (|RemoveEventMatch|_|) x =
 let (|InfoEventMatch|_|) x =
   x
     |> object
-    |> Option.bind (matchAction "info")
-    |> Option.map (fun _ -> { ACTION = "info"; })
+    |> Option.bind (matchAction (Action("info")))
+    |> Option.map (fun _ -> { ACTION = Action("info"); })
```
Comment (jgrund) [inline IML/DeviceScannerDaemon/EventTypes.fs]: May as well lift all these up as lets:

```
let add = Action("add")
let remove = Action("remove")
let info = Action("info")

File: IML/DeviceScannerDaemon/TestFixtures.fs [hunk 4]
```diff
@@ -40,7 +56,10 @@ let addObj =  toJson """
   "SUBSYSTEM": "block",
   "TAGS": ":systemd:",
   "USEC_INITIALIZED": "842",
-  "IML_SIZE": "81784832"
+  "IML_SIZE": "81784832",
+  "IML_SCSI_80": "80",
+  "IML_SCSI_83": "83",
+  "IML_IS_RO": "isRo"
```
Comment (jgrund) [inline IML/DeviceScannerDaemon/TestFixtures.fs]: this should be either "1" or "0"

File: IML/DeviceScannerDaemon/EventTypes.fs [hunk 5]
```diff
@@ -162,27 +189,27 @@ let extractAddEvent x =
...
-    IML_SIZE = parseImlSize x |> Option.bind(emptyStrToNone);
-    IML_IS_RO = imlRo;
-    IML_SCSI_80 = parseImlScsi80 x |> trimOpt;
-    IML_SCSI_83 = parseImlScsi83 x |> trimOpt;
+    ID_FS_TYPE = parseIdFsType x |> Option.bind(fun (IdFsType(x):IdFsType) -> emptyStrToNone x) |> Option.map IdFsType;
+    ID_PART_ENTRY_NUMBER = parseIdPartEntryNumber x
+    IML_SIZE = parseImlSize x |> Option.bind(fun (ImlSize(x):ImlSize) -> emptyStrToNone x) |> Option.map ImlSize;
+    IML_IS_RO = parseImlRo x
+    IML_SCSI_80 = parseImlScsi80 x |> Option.bind(fun (ImlScsi80(x):ImlScsi80) -> trimOpt(Some(x))) |> Option.map ImlScsi80;
+    IML_SCSI_83 = parseImlScsi83 x |> Option.bind(fun (ImlScsi83(x):ImlScsi83) -> trimOpt(Some(x))) |> Option.map ImlScsi83;
```
Comment (jgrund) [inline IML/DeviceScannerDaemon/EventTypes.fs]: I wonder if there is a cleaner way to do this.

File: IML/DeviceScannerDaemon/EventTypes.fs (hunk 1 above)
Comment (johnsonw) [reply IML/DeviceScannerDaemon/EventTypes.fs]: done

File: IML/DeviceScannerDaemon/EventTypes.fs (hunk 2 above)
Comment (johnsonw) [reply IML/DeviceScannerDaemon/EventTypes.fs]: Updated to:
```
let private intToIdPartEntryNumber = Option.map int >> Option.map IdPartEntryNumber
let private parseIdPartEntryNumber x = findOrNone "ID_PART_ENTRY_NUMBER" x |> intToIdPartEntryNumber
```

File: IML/DeviceScannerDaemon/EventTypes.fs (hunk 5 above)
Comment (johnsonw) [reply IML/DeviceScannerDaemon/EventTypes.fs]: I'll dig into this some more. I guess one option is to create a trimOpt that accepts an ImlScsi80 or ImlScsi83. That feels a bit awkward to me though.

File: IML/DeviceScannerDaemon/EventTypes.fs (hunk 3 above)
Comment (johnsonw) [reply IML/DeviceScannerDaemon/EventTypes.fs]: Updated

File: IML/DeviceScannerDaemon/TestFixtures.fs (hunk 4 above)
Comment (johnsonw) [reply IML/DeviceScannerDaemon/TestFixtures.fs]: Updated to "1"

File: IML/DeviceScannerDaemon/EventTypes.fs (hunk 5 above)
Comment (jgrund) [reply IML/DeviceScannerDaemon/EventTypes.fs]: You don't need the opt actually.

```
    IML_SCSI_80 = parseImlScsi80 x |> Option.map(fun (ImlScsi80(x)) -> trim(x) |> ImlScsi80);
    IML_SCSI_83 = parseImlScsi83 x |> Option.map(fun (ImlScsi83(x)) -> trim(x) |> ImlScsi83);
```

will do the trick.

What I was thinking of is methods on DU's Like `Option` has. But It doesn't matter at this juncture.

File: IML/DeviceScannerDaemon/EventTypes.fs [hunk 6]
```diff
@@ -128,18 +134,18 @@ let private findOrFail (key:string) x =
...
 )
 
 let private emptyStrToNone x = if x = "" then None else Some(x)
 
-let private trimOpt = Option.map(fun (x:string) -> x.Trim())
-
 let private findOrNone key x =
   x |> Map.tryFind key |> Option.bind str
 
+let private intToIdPartEntryNumber = Option.map int >> Option.map IdPartEntryNumber
```
Comment (jgrund) [inline IML/DeviceScannerDaemon/EventTypes.fs]: There is no need to map twice, the compose can be moved inside the map.

File: IML/DeviceScannerDaemon/EventTypesTest.fs [hunk 7]
```diff
@@ -56,7 +56,7 @@ test "Matching Events" <| fun () ->
...
...ag\":0,\"data\":\"8:16\"}]; [\"ID_PART_ENTRY_NUMBER\",{\"tag\":0,\"data\":\"1\"}]; [\"ID_PART_ENTRY_OFFSET\",{\"tag\":0,\"data\":\"2048\"}]; [\"ID_PART_ENTRY_SCHEME\",{\"tag\":0,\"data\":\"dos\"}]; [\"ID_PART_ENTRY_SIZE\",{\"tag\":0,\"data\":\"2048\"}]; [\"ID_PART_ENTRY_TYPE\",{\"tag\":0,\"data\":\"0x83\"}]; [\"ID_PART_TABLE_TYPE\",{\"tag\":0,\"data\":\"dos\"}]; [\"ID_PATH\",{\"tag\":0,\"data\":\"pci-0000:00:01.1-ata-1.0\"}]; [\"ID_PATH_TAG\",{\"tag\":0,\"data\":\"pci-0000_00_01_1-ata-1_0\"}]; [\"ID_REVISION\",{\"tag\":0,\"data\":\"1.0\"}]; [\"ID_SERIAL\",{\"tag\":0,\"data\":\"VBOX_HARDDISK_VB304a0a0f-15e93f07\"}]; [\"ID_SERIAL_SHORT\",{\"tag\":0,\"data\":\"VB304a0a0f-15e93f07\"}]; [\"ID_TYPE\",{\"tag\":0,\"data\":\"disk\"}]; [\"IML_IS_RO\",{\"tag\":0,\"data\":\"isRo\"}]; [\"IML_SCSI_80\",{\"tag\":0,\"data\":\"80\"}]; [\"IML_SCSI_83\",{\"tag\":0,\"data\":\"83\"}]; [\"IML_SIZE\",{\"tag\":0,\"data\":\"81784832\"}]; [\"MAJOR\",{\"tag\":0,\"data\":\"8\"}]; [\"MINOR\",{\"tag\":0,\"data\":\"17\"}]; [\"SEQNUM\",{\"tag\":0,\"data\":\"1566\"}]; [\"SUBSYSTEM\",{\"tag\":0,\"data\":\"block\"}]; [\"TAGS\",{\"tag\":0,\"data\":\":systemd:\"}]; [\"USEC_INITIALIZED\",{\"tag\":0,\"data\":\"842\"}]]"
```
Comment (jgrund) [inline IML/DeviceScannerDaemon/EventTypesTest.fs]: What changed here, it's hard to tell.

File: IML/DeviceScannerDaemon/HandlersTest.fs [hunk 8]
```diff
@@ -47,7 +47,7 @@ testList "Data Handler" [
 
         handler (toJson """{ "ACTION": "info" }""")
 
-        ``end`` <?> Some("{\"/devices/pci0000:00/0000:00:01.1/ata1/host1/target1:0:0/1:0:0:0/block/sdb/sdb1\":{\"ACTION\":\"add\",\"MAJOR\":\"8\",\"MINOR\":\"17\",\"DEVLINKS\":\"/dev/disk/by-id/ata-VBOX_HARDDISK_VB304a0a0f-15e93f07-part1 /dev/disk/by-path/pci-0000:00:01.1-ata-1.0-part1\",\"PATHS\":[\"/dev/sdb1\",\"/dev/disk/by-id/ata-VBOX_HARDDISK_VB304a0a0f-15e93f07-part1\",\"/dev/disk/by-path/pci-0000:00:01.1-ata-1.0-part1\"],\"DEVNAME\":\"/dev/sdb1\",\"DEVPATH\":\"/devices/pci0000:00/0000:00:01.1/ata1/host1/target1:0:0/1:0:0:0/block/sdb/sdb1\",\"DEVTYPE\":\"partition\",\"ID_VENDOR\":null,\"ID_MODEL\":\"VBOX_HARDDISK\",\"ID_SERIAL\":\"VBOX_HARDDISK_VB304a0a0f-15e93f07\",\"ID_FS_TYPE\":\"LVM2_member\",\"ID_PART_ENTRY_NUMBER\":1,\"IML_SIZE\":\"81784832\",\"IML_SCSI_80\":\"80\",\"IML_SCSI_83\":\"83\",\"IML_IS_RO\":false}}")
```
Comment (jgrund) [inline IML/DeviceScannerDaemon/HandlersTest.fs]: What changed here, it's hard to tell.

File: IML/DeviceScannerDaemon/EventTypes.fs (hunk 6 above)
Comment (johnsonw) [reply IML/DeviceScannerDaemon/EventTypes.fs]: So perhaps something like this:
```
let private intToIdPartEntryNumber = Option.map (int >> IdPartEntryNumber)
```
?

Comment (jgrund) [reply IML/DeviceScannerDaemon/EventTypes.fs]: yes.

**Pull request merged**